psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -59,82 +60,89 @@ public double optimize(TermStore baseTermStore) {
-            objective = computeObjective(termStore, variableValues);
-            totalTime += end - start;
+            converged = breakOptimization(objective, oldObjective, termCount);
 
-            if (log.isTraceEnabled()) {
-                log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}",
-                        iteration, objective, (end - start), totalTime);
+            // Keep track of the old variables for a deferred objective computation.
+            if (oldVariableValues == null) {
+                oldVariableValues = Arrays.copyOf(termStore.getVariableValues(), termStore.getVariableValues().length);
+                oldObjective = Double.POSITIVE_INFINITY;
+            } else {
+                System.arraycopy(termStore.getVariableValues(), 0, oldVariableValues, 0, oldVariableValues.length);
+                oldObjective = objective;
             }
 
-            iteration++;
-            termStore.iterationComplete();
+            long end = System.currentTimeMillis();
+            totalTime += end - start;
 
-            if (breakOptimization(iteration, objective, oldObjective)) {
-                break;
+            if (iteration > 1) {
+                if (log.isTraceEnabled()) {
+                    log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Iteration Time: {}, Total Optimization Time: {}",
+                            iteration - 1, objective, objective / termCount, (end - start), totalTime);
+                }
             }
         }
 
-        log.info("Optimization completed in {} iterations. Objective: {}, Total Optimiztion Time: {}",
-                iteration - 1, objective, totalTime);
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Final Objective: {}, Final Normalized Objective: {}, Total Optimization Time: {}", objective, objective / termCount, totalTime);
+        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termCount);
 
         termStore.syncAtoms();
 
-        return objective;
+        return objective / termCount;
     }
 
-    private boolean breakOptimization(int iteration, double objective, double oldObjective) {
-        // Always break when the allocated iterations is up.
-        if (iteration > (int)(maxIterations * budget)) {
-            return true;
-        }
-
+    private boolean breakOptimization(double objective, double oldObjective, long termCount) {
         // Run through the maximum number of iterations.
         if (runFullIterations) {
             return false;
         }
 
         // Break if the objective has not changed.
-        if (objectiveBreak && MathUtils.equals(objective, oldObjective, tolerance)) {
+        if (objectiveBreak && MathUtils.equals(objective / termCount, oldObjective / termCount, tolerance)) {
```

Comment:
again, does this want to be normalized now?

Reply:
I argue yes since if the objective is normalized by termCount here, then the tolerance does not need to be scaled with respect to the number of terms; the tolerance has the same interpretation for models with a different number of terms.
Setting the default value for tolerance would make sense for more PSL models and in settings such as online, one does not need to adjust tolerance before and after model modifications.